### ISSUE TITLE
Failed to process query. Incorrect syntax. 

### ISSUE THREAD

**Issue by aloscha** [Starcounter/Starcounter#3716]
I have two classes, example:
`public class OrderItem : Something {`
`    public Order Order;`
`    public int Quantity;`
`}`
And
`public class Order : Something{ }`

I would like to find all OrderItem from current order:
`Db.SQL("SELECT r FROM Simplified.Ring6.OrderItem r WHERE r.Order.Key = SomeKey ")`

But I receive Failed to process query. Incorrect syntax.

**Comment by k-rus** [comment]
Thank you for reporting. `ORDER` is reserved keyword, so it needs to be covered with double quotes to make it to be identifier. See [the list of keywords](http://starcounter.io/guides/sql/reserved-words/) and [page about identifiers](http://starcounter.io/guides/sql/identifiers/)

**Comment by per-samuelsson** [comment]
We have an old working item to fix this, #1989, that could possibly provide some more info too.

**Comment by warpech** [comment]
Is there a chance to improve the error message so that it is a little bit more pointing to the right direction? We have had exactly the same problem so many times already. Yet, while trying to help @aloscha, I still missed to connect the dots that the word `ORDER` causes the problem.

**Comment by warpech** [comment]
What is more confising - @aloscha reports that such query works ok in C#. Only Administrator is the problem.

**Comment by per-samuelsson** [comment]
> Is there a chance to improve the error message so that it is a little bit more pointing to the right direction? 

I'll let @k-rus answer that, but my impression was that the idea was instead to release "ORDER" from being a reserved word, according to #1989. That would be even better IMO.

> What is more confising - @aloscha reports that such query works ok in C#. Only Administrator is the problem.

Agree, that seem strange and clearly a source of confusion.

**Comment by k-rus** [comment]
> We have an old working item to fix this, #1989, that could possibly provide some more info too.

The working item is about new query processor, while the error is in the old one, so #1989 is not related.



> What is more confising - @aloscha reports that such query works ok in C#. Only Administrator is the problem.

According to [the initial report](https://github.com/Starcounter/Starcounter/issues/3716#issue-161467244) from @aloscha, he gets the error in C#, since he specifies use of `Db.SQL`. See:



> I would like to find all OrderItem from current order:
> Db.SQL("SELECT r FROM Simplified.Ring6.OrderItem r WHERE r.Order.Key = SomeKey ")
> 
> But I receive Failed to process query. Incorrect syntax.



> Is there a chance to improve the error message so that it is a little bit more pointing to the right direction?
> I'll let @k-rus answer

The prolog parser, which is used in the old query processor, does not report details. Thus it is difficult to produce meaningful error. One approach is to write a parser on top to produce more meaningful error, but it seems to me to be overkill. I would rather spend resources on actual work with new query processor.

I can try to debug the old query processor to confirm that no meaningful information is reported.

**Comment by per-samuelsson** [comment]
>  I would rather spend resources on actual work with new query processor.

Yeah, it's a challenge trying to find the right balance. We really want focus to be on the new processor, messing as little as possible with the old one.

**Comment by warpech** [comment]
Today, another person was hit by this problem:

![2016-12-20 18_13_09-kitchensink - microsoft visual studio](https://cloud.githubusercontent.com/assets/566463/21360396/04cef69c-c6e0-11e6-89e4-295a02c6ca84.png)

**Comment by aloscha** [comment]
@warpech to perfect solution is using linq there. But yes it should be escaped somehow.

**Comment by andan342** [comment]
In the new parser, both mentioned queries
```
SELECT r FROM Simplified.Ring6.OrderItem r WHERE r.Order.Key = F

SELECT p FROM Person p ORDER BY p.Order ASC
```
are parsed correctly - there should be no syntax problem since the #2081fix - i.e. August 2014.

So I guess the problems reported here correspond to the Prolog-based parser.
